### PR TITLE
Add annotation convention for serialization

### PR DESCRIPTION
**Pull request by simonz-bq**

### CODE REVIEW

**Comment by justing-bq** [comment]
Can you add a test for this change? Or is it already handled by the existing tests?

File: Amazon.Ion.ObjectMapper/TypeAnnotator.cs [hunk 1]
```diff
@@ -45,29 +71,29 @@ public class DefaultTypeAnnotator : TypeAnnotator
...
                 annotateType = new IonAnnotateType();
             }
 
             if (annotateType != null)
             {
-                var prefix = annotateType.Prefix != null ? annotateType.Prefix : options.TypeAnnotationPrefix.Apply(type);
-                var name = annotateType.Name != null ? annotateType.Name : type.Name;
-                writer.AddTypeAnnotation(prefix + "." + name);
+                annotateType.Prefix = annotateType.Prefix != null ? annotateType.Prefix : options.TypeAnnotationPrefix.Apply(type);
+                annotateType.Name = annotateType.Name != null ? annotateType.Name : options.TypeAnnotationName.Apply(type);
```
[the remark this answers is not in the document]
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper/TypeAnnotator.cs:83]: Or alternatively:

```c#
annotateType.Prefix ??= options.TypeAnnotationPrefix.Apply(type);
annotateType.Name ??= options.TypeAnnotationName.Apply(type);
```

[the remark this answers is not in the document]
Comment (simonz-bq) [reply Amazon.Ion.ObjectMapper/TypeAnnotator.cs]: Done.

**Comment by simonz-bq** [comment]
> Can you add a test for this change? Or is it already handled by the existing tests?

It's covered by the existing tests.

File: Amazon.Ion.ObjectMapper/TypeAnnotator.cs [hunk 2]
```diff
@@ -35,6 +35,32 @@ public string Apply(Type type)
         }
     }
 
+    public interface TypeAnnotationName
+    {
+        public string Apply(Type type);
```
Comment (justing-bq) [inline Amazon.Ion.ObjectMapper/TypeAnnotator.cs:40]: Can these Apply() methods be static?

File: Amazon.Ion.ObjectMapper/TypeAnnotator.cs [hunk 3]
```diff
@@ -35,6 +35,32 @@ public string Apply(Type type)
...
+    {
+        public string Apply(Type type)
+        {
+            return type.Name;
+        }
+    }
+
+    public interface AnnotationConvention
+    {
+        public string Apply(IonAnnotateType annotateType, Type type);
```
Comment (justing-bq) [inline Amazon.Ion.ObjectMapper/TypeAnnotator.cs:53]: Likewise. Can these be static?

File: Amazon.Ion.ObjectMapper/TypeAnnotator.cs (hunk 2 above)
Comment (simonz-bq) [reply Amazon.Ion.ObjectMapper/TypeAnnotator.cs:40]: No, because it's an interface method.

File: Amazon.Ion.ObjectMapper/TypeAnnotator.cs (hunk 3 above)
Comment (simonz-bq) [reply Amazon.Ion.ObjectMapper/TypeAnnotator.cs:53]: Ditto.

File: Amazon.Ion.ObjectMapper/IonSerializer.cs [hunk 4]
```diff
@@ -101,24 +101,40 @@ private string FullName(string typeName, string assemblyName)
     public class IonSerializationOptions
     {
         public IonPropertyNamingConvention NamingConvention { get; init; } = new CamelCaseNamingConvention();
+
```
Comment (kylepbit) [inline Amazon.Ion.ObjectMapper/IonSerializer.cs:104]: Why all the additions of whitespace - please revert for now until we apply styling.

File: Amazon.Ion.ObjectMapper/IonSerializer.cs [hunk 5]
```diff
@@ -233,19 +249,19 @@ public void Serialize<T>(IIonWriter writer, T item)
                 return;
             }
 
-            if (item is System.Collections.IList) 
+            if (item is System.Collections.IList)
```
Comment (kylepbit) [inline Amazon.Ion.ObjectMapper/IonSerializer.cs:252]: I understand what's being done and why, but can we revert the changes for trailing whitespace and put them in a separate PR since it makes it very hard to see what's actually changed.

**Review by kylepbit — CHANGES_REQUESTED** [review]

File: Amazon.Ion.ObjectMapper/IonSerializer.cs (hunk 4 above)
Comment (simonz-bq) [reply Amazon.Ion.ObjectMapper/IonSerializer.cs]: Fixed

File: Amazon.Ion.ObjectMapper/IonSerializer.cs (hunk 5 above)
Comment (simonz-bq) [reply Amazon.Ion.ObjectMapper/IonSerializer.cs]: Fixed

**Review by alpian — APPROVED** [review]

**Comment by alpian** [comment]
Nice work! :)
